bus-mapping/src/rpc.rs

```diff
@@ -0,0 +1,185 @@
+    /// Serializes a BlockNumber as a [`Value`](serde_json::Value) to be able to throw it into a JSON-RPC request.
+    pub fn serialize(self) -> serde_json::Value {
+        match self {
+            BlockNumber::Num(num) => serialize(&num),
+            BlockNumber::Earliest => serialize(&"earliest"),
+            BlockNumber::Latest => serialize(&"latest"),
+            BlockNumber::Pending => serialize(&"pending"),
+        }
+    }
+}
+
+/// Placeholder structure designed to contain the methods that the BusMapping needs in order to enable Geth queries.
+pub struct GethClient<P: JsonRpcClient>(P);
+
+impl<P: JsonRpcClient> GethClient<P> {
+    /// Generates a new `GethClient` instance.
+    pub fn new(provider: P) -> Self {
+        Self(provider)
+    }
+
+    /// Calls `eth_getBlockByHash` via JSON-RPC returning a [`Block`] returning all the block information including it's transaction's details.
+    pub async fn get_block_by_hash(
+        &self,
+        hash: Hash,
+    ) -> Result<Block<Transaction>, Error> {
+        let hash = serialize(&hash);
+        let flag = serialize(&true);
+        self.0
+            .request("eth_getBlockByHash", [hash, flag])
+            .await
+            .map_err(|e| Error::JSONRpcError(e.into()))
+    }
+
+    /// Calls `eth_getBlockByNumber` via JSON-RPC returning a [`Block`] returning all the block information including it's transaction's details.
+    pub async fn get_block_by_number(
+        &self,
+        block_num: BlockNumber,
+    ) -> Result<Block<Transaction>, Error> {
+        let num = block_num.serialize();
+        let flag = serialize(&true);
+        self.0
+            .request("eth_getBlockByNumber", [num, flag])
+            .await
+            .map_err(|e| Error::JSONRpcError(e.into()))
+    }
+
+    /// Calls `debug_traceBlockByHash` via JSON-RPC returning a [`Vec<GethExecTrace>`] with each GethTrace corresponding to 1 transaction of the block.
+    pub async fn trace_block_by_hash(
+        &self,
+        hash: Hash,
+    ) -> Result<Vec<GethExecTrace>, Error> {
+        let hash = serialize(&hash);
+        let resp: ResultGethExecTraces = self
+            .0
+            .request("debug_traceBlockByHash", [hash])
+            .await
+            .map_err(|e| Error::JSONRpcError(e.into()))?;
+        Ok(resp.0.into_iter().map(|step| step.result).collect())
+    }
+
+    /// Calls `debug_traceBlockByNumber` via JSON-RPC returning a [`Vec<GethExecTrace>`] with each GethTrace corresponding to 1 transaction of the block.
+    pub async fn trace_block_by_number(
+        &self,
+        block_num: BlockNumber,
+    ) -> Result<Vec<GethExecTrace>, Error> {
+        let num = block_num.serialize();
+        let resp: ResultGethExecTraces = self
+            .0
+            .request("debug_traceBlockByNumber", [num])
+            .await
+            .map_err(|e| Error::JSONRpcError(e.into()))?;
+        Ok(resp.0.into_iter().map(|step| step.result).collect())
+    }
```

Comment:
FWIW we'd be happy to upstream the Geth Tracing logic to ethers-rs, similar to the parity tracing module https://github.com/gakonst/ethers-rs/blob/master/ethers-providers/src/provider.rs#L788

Reply:
The problem is that we use a slightly different variant of the types that you have defined in your lib. 

That's why we did not make a PR adding them in `ethers` and instead just took the `Provider` backend and used it directly.

Will try to add this in a PR in ethers once I get a bit more time :)